### PR TITLE
Use `common_type` for complex `pow`

### PR DESCRIPTION
**Pull request by miscco**
Previously we would rely on our internal `__promote` function.

However, that could have surprising results, e.g. `pow(complexy<float>, int)` would return `complex<double>`

With C++23, this situation got clarified and we should use `common_type` to determine the return type.

### CODE REVIEW

File: libcudacxx/include/cuda/std/detail/libcxx/include/cmath [hunk 1]
```diff
@@ -603,7 +603,7 @@ inline _LIBCUDACXX_INLINE_VISIBILITY long double hypot(long double x, long doubl
 template <class _A1, class _A2, class _A3>
 inline _LIBCUDACXX_INLINE_VISIBILITY
   __enable_if_t<is_arithmetic<_A1>::value && is_arithmetic<_A2>::value && is_arithmetic<_A3>::value,
-                __promote<_A1, _A2, _A3>>
+                __promote_t<_A1, _A2, _A3>>
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/cmath:606]: Drive-by-fix

**Comment by bernhardmgruber** [comment]
I don' quite see it in the code, but what happens with `pow(complex<__half>, complex<bfloat16>)`? If I understand [P1467](https://wg21.link/p1467r9) correctly, `float16_t` and `bfloat16_t` have unordered conversion ranks, so combining them in arithmetic should be a compilation error. The implementation proposed here looks like it would promote to `complex<float>`.

File: libcudacxx/include/cuda/std/__type_traits/common_type.h [hunk 2]
```diff
@@ -127,7 +127,44 @@ _LIBCUDACXX_INLINE_VAR constexpr bool __has_common_type = false;
...
+struct common_type<T, __nv_bfloat16> : common_type<T, float>
+{};
+
+// __half and __nvbfloat16 have unordered conversion rank
+template <>
+struct common_type<__half, __nv_bfloat16>
+{};
+template <>
+struct common_type<__nv_bfloat16, __half>
+{};
```
Comment (bernhardmgruber) [inline libcudacxx/include/cuda/std/__type_traits/common_type.h:166]: Question: Does `_LIBCUDACXX_HAS_NVBF16` imply that `__half` is available? If not, then you may need to guard this code also by `_LIBCUDACXX_HAS_NVFP16`.

**Review by bernhardmgruber — APPROVED** [review]

File: libcudacxx/include/cuda/std/__type_traits/common_type.h (hunk 2 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/__type_traits/common_type.h:166]: Yes, the nvbfloat16 header depends on the half header
